openquake/commonlib/calculators/event_based.py

```diff
@@ -192,43 +186,77 @@ def compute_ruptures(sources, sitecol, monitor):
         if s_sites is None:
             continue
 
-        # the dictionary `ses_num_occ` contains [(ses, num_occurrences)]
-        # for each occurring rupture for each ses in the ses collection
-        ses_num_occ = collections.defaultdict(list)
-        # generating ruptures for the given source
-        for rup_no, rup in enumerate(src.iter_ruptures(), 1):
-            rup.rup_no = rup_no
-            for ses_idx in all_ses:
+        num_occ_by_rup = sample_ruptures(
+            src, oq.ses_per_logic_tree_path, info, rnd)
+        # NB: the number of occurrences is very low, << 1, so it is
+        # more efficient to filter only the ruptures that occur, i.e.
+        # to call sample_ruptures *before* the filtering
+
+        for rup, rups in filter_ruptures(
+                num_occ_by_rup, s_sites, oq.maximum_distance,
+                sitecol, rnd, src):
+            sesruptures.extend(rups)
+
+    return {trt_model_id: sesruptures}
+
+
+def sample_ruptures(src, num_ses, info, rnd):
+    """
+    Sample the ruptures contained in the given source.
+
+    :param src: a hazardlib source object
+    :param num_ses: the number of Stochastic Event Sets to generate
+    :param rnd: an instance of :class:`random.Random`
```

Comment:
please swap params `rnd` and `info`, making the order consistent with the signature